src/System.Management.Automation/engine/MshObjectIConvertible.cs

```diff
@@ -0,0 +1,104 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#nullable enable
+
+using System.Diagnostics.CodeAnalysis;
+
+namespace System.Management.Automation
+{
+    /// <summary>
+    /// <see cref="PSObject"/> implements <see cref="System.IConvertible"/> interface for its base object.
+    /// </summary>
+    /// <remarks>
+    /// If <see cref="PSObject"/> wraps a value type implementing <see cref="System.IConvertible"/>
+    /// it provides that functionality too.
+    /// </remarks>
+    /// <exception cref="System.InvalidCastException">The base object's implementation cannot convert and throws.</exception>
+    public partial class PSObject : IConvertible
+    {
+        /// <summary>
+        /// Returns the type code of the base object wrapped by the <see cref="PSObject"/>.
+        /// </summary>
+        /// <remarks>
+        /// If the base object is <see cref="PSObject"/> the method returns <see cref="TypeCode.Object"/>
+        /// i.e. it does not implement <see cref="System.IConvertible"/>.
+        /// Otherwise, the result is the type code of the base object,
+        /// as determined by the base object's implementation of <see cref="System.IConvertible"/>.
+        /// </remarks>
+        /// <returns>Returns <see cref="System.TypeCode"/> of the PSObject.</returns>
+        public TypeCode GetTypeCode()
+        {
+            object obj = PSObject.Base(this);
+            return obj is PSObject ? TypeCode.Object : Convert.GetTypeCode(obj);
+        }
+
+        /// <inheritdoc/>
+        public bool ToBoolean(IFormatProvider? provider) => GetIConvertibleOrThrow(this).ToBoolean(provider);
```

Comment:
All members should be explicit definitions, like `bool IConvertible.ToBoolean(IFormatProvider? provider) => ...`

However, changing to explicit definitions does address the exception we saw. You get the same error.